### PR TITLE
Variable Granularity support for data collection

### LINKED ISSUE #6254: Netdata variable granularity support in dbengine
##### Feature idea summary
Netdata variable granularity support in dbengine.
##### Expected behavior
dbengine API will need to implement API calls than enable variable granularity of data collection frequency of metrics collected by netdata or other API users (see #6255).

### LINKED ISSUE #6255: Netdata variable granularity support in netdata daemon - with basic unit testing
##### Feature idea summary
Netdata variable granularity support in netdata daemon.
##### Expected behavior
Allow users to change data collection frequency of metrics without:
- deleting older chart data
- invalidating existing alarm data
- affecting streaming
- invalidating web server data queries of older data
- affecting backends

### LINKED ISSUE #6487: Functional testing of variable granularity in data collection
##### Feature idea summary
Functional testing of variable granularity in data collection. Tests that the feature #6255 produces correct results when netdata is queried about metric data.

### LINKED ISSUE #6573: database query returns 0 points when it should return 1
##### Bug report summary
Database query results are wrong on http://frankfurt.my-netdata.io/. 
##### OS / Environment
Debian 
##### Netdata version (ouput of `netdata -V`)
v1.14 v1.16
##### Component Name
database
##### Steps To Reproduce
```
2019-07-31 12:44:53: netdata INFO  : HEALTH : INTERNAL ERROR: rrd2rrdr() on netdata_cloud_pages.sso_agent_responses update every 1 with unaligned grouping sum (group: 3600, resampling_time: 0, resampling_group: 1), after (got: 1564566293, want: 1564566293, req: 1564566292, db: 1564566203), before (got: 1564569892, want: 1564569892, req: 1564569892, db: 1564569892), duration (got: 3600, want: 3600, req: 3600, db: 3690), points (got: 0, want: 1, req: 1, db: 172956), got 'points' is not wanted 'points'
```
```
2019-07-31 12:38:19: netdata INFO  : WEB_SERVER[static1] : INTERNAL ERROR: rrd2rrdr() on dockerhub.pulls_sum update every 5 with aligned grouping incremental_sum (group: 657, resampling_time: 0, resampling_group: 1), after (got: 18446744073709548336, want: 1564566660, req: 1564566
003, db: 1564566210), before (got: 0, want: 1564569288, req: 1564569288, db: 1564569495), duration (got: 3285, want: 2633, req: 3285, db: 3290), points (got: 0, want: 1, req: 1, db: 36764), got 'after' is not wanted 'after'
```
```
2019-07-31 13:38:40: netdata INFO  : HEALTH : INTERNAL ERROR: rrd2rrdr() on netdata_cloud_pages.home_responses update every 1 with unaligned grouping sum (group: 3600, resampling_time: 0, resampling_group: 1), after (got: 1564569520, want: 1564569520, req: 1564569519, db: 1564568570), before (got: 1564573119, want: 1564573119, req: 1564573119, db: 1564573119), duration (got: 3600, want: 3600, req: 3600, db: 4550), points (got: 0, want: 1, req: 1, db: 172956), got 'points' is not wanted 'points'
```
##### Expected behavior
Points returned should be 1 instead of 0. Warnings should not be produced.

### PR DESCRIPTION
**Pull request by mfundul**
##### Summary
Fixes #6254 
Fixes #6255
Fixes #6487
Fixes #6573
Variable Granularity support for data collection in netdata.
##### Component Name
daemon
database
web
backends
##### Additional Information
Allow users to change data collection frequency of metrics without:
- deleting older chart data
- invalidating existing alarm data
- affecting streaming
- invalidating web server data queries of older data
- affecting backends

### CODE REVIEW

**Comment by mfundul** [comment]
To test this one can stop the netdata service, change the `update_every` interval of some charts, and at a later point in time enable the service again.

The focus for testing should be that there are no bogus metric data being returned when the database is queried.

This can be tested by observing dashboard queries, web API queries, health queries, backend queries etc.

**Comment by thiagoftsm** [comment]
After to run during almost 3 hours the netdata stopped to answer during 2 minutes, but it returned
until now I realized 1,411,917 requests to the database (edited) 
Now, it stopped again in the stress test
Another thing that I notice and are letting me worried is when I try to access the dashboard. I opened the firefox in "private window" to avoid cache, and I am getting data only with timestamp inside the range 5:34:00 PM and 5:40:00 PM UTC, but right now UTC time is 8:00 PM (edited) 
The valgrind is also giving me a lot of messages like this:
==3449== Warning: set address range perms: large range [0x1c717010, 0x33c88881) (defined)
==3449== Warning: set address range perms: large range [0x1c717010, 0x33c88881) (defined)
==3449== Warning: set address range perms: large range [0x1c717010, 0x33c88881) (defined)
==3449== Warning: set address range perms: large range [0x1c717010, 0x33c88881) (defined)
==3449== Warning: set address range perms: large range [0x1c717010, 0x33c88881) (defined)
==3449== Warning: set address range perms: large range [0x1c717010, 0x33c88881) (defined)
==3449== Warning: set address range perms: large range [0x39606000, 0x5afbc000) (defined)
==3449== Warning: set address range perms: large range [0x39606000, 0x5afbc000) (noaccess)
==3449== Warning: set address range perms: large range [0x1c717000, 0x33c89000) (noaccess)
Finally I am having also valgrind erros like :
==3449== Warning: set address range perms: large range [0x39606000, 0x5afbc000) (noaccess)
==3449== Syscall param epoll_pwait(sigmask) points to unaddressable byte(s)
==3449==    at 0x597460E: ??? (in /lib64/libc-2.29.so)
==3449==    by 0x2C3000000001036: ???
==3449==    by 0x5982C85: ??? (in /lib64/libc-2.29.so)
==3449==    by 0x2CC: ???
==3449==    by 0x6A85E4F: ???
==3449==    by 0x6A85EA7: ???
==3449==    by 0xFFEFFD6CD: ???
==3449==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
these two groups are the most common error that I am receiving
Until now it ran 8 loops of the request script and it is also running the stress test, when it completes 10 loops, in 40 minutes,  I will stop and restart to confirm that the dashboards is only getting the charts for the first time that Netdata was started.

After I stop everything and start the test again, the dashboard is running fine this last 10 minutes, but I am receiving a lot of messages like this from valgrind:
"==10246==  Access not within mapped region at address 0xBD5D158
==10246==    at 0x441A5C: ??? (in /home/thiago/Netdata/netdata/netdata)
==10246==    by 0x4429BF: ??? (in /home/thiago/Netdata/netdata/netdata)
==10246==    by 0x50C538: ??? (in /home/thiago/Netdata/netdata/netdata)
==10246==    by 0x4EE519B: ??? (in /lib64/libpthread-2.29.so)
==10246==  If you believe this happened as a result of a stack
==10246==  overflow in your program's main thread (unlikely but
==10246==  possible), you can try to increase the size of the
==10246==  main thread stack using the --main-stacksize= flag.
==10246==
==10246== HEAP SUMMARY:
==10246==     in use at exit: 0 bytes in 0 blocks
==10246==   total heap usage: 0 allocs, 0 frees, 0 bytes allocated
==10246==
==10246== All heap blocks were freed -- no leaks are possible
==10246==
==10246== For counts of detected and suppressed errors, rerun with: -v
==10246== ERROR SUMMARY: 747 errors from 1 contexts (suppressed: 0 from 0)
"
And the number of errors is increasing
in 2 minutes the cron will start the routine of run the request.sh from tests/urls each 20 minutes

During the first run of the request.sh the Netdata dashboard stopped to show the chart update and valgrind began to show messages like :
==9943== Warning: set address range perms: large range [0x1cb67010, 0x3416b40f) (defined)
==9943== Warning: set address range perms: large range [0x1cb67010, 0x3416b40f) (defined)
I stopped the tests after 260 minutes, mainly due the problem that we are having with dashboard when there is a great number of requests going to Netdata, the other errors, this does not mean that the other errors and warnings are not important, but without any doubt this is the first that the user will see.
I wrote a long report, so I will resume here what is necessary to recreate the problem:
1 - Compile Netdata with debug symbols (edited) 
2 - Run Netdata inside valgrind
3 - Open a browser and access netdata, this will occur without any problem
4 - Put the script test/urls/request.sh to run from crontab starting each 20 minutes
5 - I also extended the level of tests using the script test/stress.sh . I believe the stress also can create the same problem described in 4 with the other script, because it looks like that his happen only when it is necessary to access a great sample of memory.
Please, let me know when I can restart the tests here. Tomorrow  I will again reserve one of my computers exclusively for this

**Comment by vlvkobal** [comment]
I've just started the testing. This is what I've got from the start:
image

My system: Arch Linux 5.2.6-arch1-1-ARCH #1 SMP PREEMPT x86_64 GNU/Linux

Netdata was built with
```
sudo CFLAGS="-O1 -ggdb -Wall -Wextra -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1" ./netdata-installer.sh --enable-plugin-nfacct --enable-plugin-freeipmi --enable-backend-kinesis --enable-backend-prometheus-remote-write --disable-go --disable-lto --dont-wait
```

I ran it using valgrind
```
sudo valgrind --leak-check=yes /usr/sbin/netdata -P /var/run/netdata/netdata.pid -D -W set global process scheduling policy keep -W set global OOM score keep
```

I have a lot of messages in the `error.log` like
```
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.cpu': discarded data collection time of 1000035 usec, replaced with 2197985 usec, reason: 'SYNC CLOCK FLAG'
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.cpu_user': discarded data collection time of 1000035 usec, replaced with 2198538 usec, reason: 'SYNC CLOCK FLAG'
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.cpu_system': discarded data collection time of 1000035 usec, replaced with 2198825 usec, reason: 'SYNC CLOCK FLAG'
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.threads': discarded data collection time of 1000035 usec, replaced with 2199157 usec, reason: 'SYNC CLOCK FLAG'

```
and
```
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.cpu': discarded data collection time of 1000035 usec, replaced with 2197985 usec, reason: 'SYNC CLOCK FLAG'
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.cpu_user': discarded data collection time of 1000035 usec, replaced with 2198538 usec, reason: 'SYNC CLOCK FLAG'
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.cpu_system': discarded data collection time of 1000035 usec, replaced with 2198825 usec, reason: 'SYNC CLOCK FLAG'
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.threads': discarded data collection time of 1000035 usec, replaced with 2199157 usec, reason: 'SYNC CLOCK FLAG'
2019-08-07 19:46:10: netdata INFO  : PLUGINSD[apps] : host 'arch-esxi', chart 'users.processes': discarded data collection time of 1000035 usec, replaced with 2199444 usec, reason: 'SYNC CLOCK FLAG'
```

valgrind reports
```
==9231== Memcheck, a memory error detector
==9231== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==9231== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==9231== Command: /usr/sbin/netdata -P /var/run/netdata/netdata.pid -D -W set global process scheduling policy keep -W set global OOM score keep
==9231==
==9231== Syscall param pwrite64(buf) points to uninitialised byte(s)
==9231==    at 0x5787408: pwrite (in /usr/lib/libpthread-2.29.so)
==9231==    by 0x4DEB04E: ??? (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x4DED197: uv_fs_write (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x1A4BEB: create_data_file (datafile.c:147)
==9231==    by 0x1A4DC6: create_new_datafile_pair (datafile.c:350)
==9231==    by 0x1A5375: init_data_files (datafile.c:394)
==9231==    by 0x1A3A1C: init_rrd_files (rrdengine.c:577)
==9231==    by 0x1A8515: rrdeng_init (rrdengineapi.c:769)
==9231==    by 0x19949D: rrdhost_create (rrdhost.c:247)
==9231==    by 0x199A09: rrd_init (rrdhost.c:496)
==9231==    by 0x13C8D9: main (main.c:1207)
==9231==  Address 0x6440031 is 49 bytes inside a block of size 4,096 alloc'd
==9231==    at 0x483AEC3: memalign (vg_replace_malloc.c:898)
==9231==    by 0x483AFF0: posix_memalign (vg_replace_malloc.c:1062)
==9231==    by 0x1A4B5B: create_data_file (datafile.c:137)
==9231==    by 0x1A4DC6: create_new_datafile_pair (datafile.c:350)
==9231==    by 0x1A5375: init_data_files (datafile.c:394)
==9231==    by 0x1A3A1C: init_rrd_files (rrdengine.c:577)
==9231==    by 0x1A8515: rrdeng_init (rrdengineapi.c:769)
==9231==    by 0x19949D: rrdhost_create (rrdhost.c:247)
==9231==    by 0x199A09: rrd_init (rrdhost.c:496)
==9231==    by 0x13C8D9: main (main.c:1207)
==9231==
==9231== Syscall param pwrite64(buf) points to uninitialised byte(s)
==9231==    at 0x5787408: pwrite (in /usr/lib/libpthread-2.29.so)
==9231==    by 0x4DEB04E: ??? (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x4DED197: uv_fs_write (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x1A5C94: create_journal_file (journalfile.c:195)
==9231==    by 0x1A4E7D: create_new_datafile_pair (datafile.c:361)
==9231==    by 0x1A5375: init_data_files (datafile.c:394)
==9231==    by 0x1A3A1C: init_rrd_files (rrdengine.c:577)
==9231==    by 0x1A8515: rrdeng_init (rrdengineapi.c:769)
==9231==    by 0x19949D: rrdhost_create (rrdhost.c:247)
==9231==    by 0x199A09: rrd_init (rrdhost.c:496)
==9231==    by 0x13C8D9: main (main.c:1207)
==9231==  Address 0x6442030 is 48 bytes inside a block of size 4,096 alloc'd
==9231==    at 0x483AEC3: memalign (vg_replace_malloc.c:898)
==9231==    by 0x483AFF0: posix_memalign (vg_replace_malloc.c:1062)
==9231==    by 0x1A5C09: create_journal_file (journalfile.c:186)
==9231==    by 0x1A4E7D: create_new_datafile_pair (datafile.c:361)
==9231==    by 0x1A5375: init_data_files (datafile.c:394)
==9231==    by 0x1A3A1C: init_rrd_files (rrdengine.c:577)
==9231==    by 0x1A8515: rrdeng_init (rrdengineapi.c:769)
==9231==    by 0x19949D: rrdhost_create (rrdhost.c:247)
==9231==    by 0x199A09: rrd_init (rrdhost.c:496)
==9231==    by 0x13C8D9: main (main.c:1207)
==9231==
```
but I don't think it is relevant to the gaps on the charts.

Errors on the exit:
```
^C==9231== Thread 4:
==9231== Syscall param pwrite64(buf) points to uninitialised byte(s)
==9231==    at 0x578744F: pwrite (in /usr/lib/libpthread-2.29.so)
==9231==    by 0x4DEB04E: ??? (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x4DE3AFE: ??? (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x577D57E: start_thread (in /usr/lib/libpthread-2.29.so)
==9231==    by 0x58910E2: clone (in /usr/lib/libc-2.29.so)
==9231==  Address 0x6eff961 is 2,401 bytes inside a block of size 4,096 alloc'd
==9231==    at 0x483AEC3: memalign (vg_replace_malloc.c:898)
==9231==    by 0x483AFF0: posix_memalign (vg_replace_malloc.c:1062)
==9231==    by 0x1A57C7: wal_get_transaction_buffer (journalfile.c:83)
==9231==    by 0x1A233E: commit_data_extent (rrdengine.c:218)
==9231==    by 0x1A3677: do_commit_transaction (rrdengine.c:242)
==9231==    by 0x1A3677: do_flush_pages (rrdengine.c:456)
==9231==    by 0x1A38FB: timer_cb (rrdengine.c:662)
==9231==    by 0x4DE47EF: uv__run_timers (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x4DE7C73: uv_run (in /usr/lib/libuv.so.1.0.0)
==9231==    by 0x1A4115: rrdeng_worker (rrdengine.c:725)
==9231==    by 0x577D57E: start_thread (in /usr/lib/libpthread-2.29.so)
==9231==    by 0x58910E2: clone (in /usr/lib/libc-2.29.so)
==9231==
==9231==
==9231== HEAP SUMMARY:
==9231==     in use at exit: 2,297,806 bytes in 57,666 blocks
==9231==   total heap usage: 554,385 allocs, 496,719 frees, 1,275,950,927 bytes allocated
==9231==
==9231== Thread 1:
==9231== 55 (48 direct, 7 indirect) bytes in 1 blocks are definitely lost in loss record 8,547 of 13,906
==9231==    at 0x483AB65: calloc (vg_replace_malloc.c:752)
==9231==    by 0x1536AD: callocz (libnetdata.c:182)
==9231==    by 0x145CD5: buffer_create (buffer.c:362)
==9231==    by 0x178CB5: backends_main (backends.c:454)
==9231==    by 0x164332: thread_start (threads.c:126)
==9231==    by 0x577D57E: start_thread (in /usr/lib/libpthread-2.29.so)
==9231==    by 0x58910E2: clone (in /usr/lib/libc-2.29.so)
==9231==
==9231== LEAK SUMMARY:
==9231==    definitely lost: 48 bytes in 1 blocks
==9231==    indirectly lost: 7 bytes in 1 blocks
==9231==      possibly lost: 0 bytes in 0 blocks
==9231==    still reachable: 2,297,751 bytes in 57,664 blocks
==9231==         suppressed: 0 bytes in 0 blocks
==9231== Reachable blocks (those to which a pointer was found) are not shown.
==9231== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==9231==
==9231== For counts of detected and suppressed errors, rerun with: -v
==9231== Use --track-origins=yes to see where uninitialised values come from
==9231== ERROR SUMMARY: 27 errors from 4 contexts (suppressed: 0 from 0)
==9231== could not unlink /tmp/vgdb-pipe-from-vgdb-to-9231-by-root-on-???
==9231== could not unlink /tmp/vgdb-pipe-to-vgdb-from-9231-by-root-on-???
==9231== could not unlink /tmp/vgdb-pipe-shared-mem-vgdb-9231-by-root-on-???
```

**Comment by vlvkobal** [comment]
My configuration:
```
[global]
        memory mode = dbengine
        enable zero metrics = yes

[backend]
        enabled = yes
        type = mongodb
```

**Comment by vlvkobal** [comment]
Without valgrind the charts are smooth.

**Comment by mfundul** [comment]
The valgrind warning were always like this and are false positives.

**Comment by mfundul** [comment]
I will have to see about this:
```
discarded data collection time of 1000035 usec, replaced with 2197985 usec, reason: 'SYNC CLOCK FLAG'
```
but I suspect it is not related to this PR.

**Comment by thiagoftsm** [comment]
> Without valgrind the charts are smooth.

I notice the same thing here, but I assumed it was and created by valgrind.

**Comment by thiagoftsm** [comment]
> 'SYNC CLOCK FLAG'

To complemente the information, I do not have any log with this message until now, I am already running the test  in the last than 6 hours, I executed 9,157,660 database queries and I received more than 7200 messages from an alarm that I am using to test this PR.

File: web/api/queries/query.c [hunk 1]
```diff
@@ -403,116 +403,254 @@ static inline void do_dimension(
...
+
+        // store the value
+        value = r->internal.grouping_flush(r, rrdr_value_options_ptr);
+        r->v[rrdr_line * r->d + dim_id_in_rrdr] = value;
+
+        if(likely(points_added || dim_id_in_rrdr)) {
+            // find the min/max across all dimensions
+
+            if(unlikely(value < min)) min = value;
+            if(unlikely(value > max)) max = value;
```
Comment (cosmix) [inline web/api/queries/query.c]: I'm not sure there's any point in using `unlikely` here.

Comment (mfundul) [reply web/api/queries/query.c]: It's just Costa's copy-pasta code I moved around. Just following the logic "if it ain't broke don't fix it".

**Pull request merged**